vendor/bat-native-ads/src/bat/ads/internal/ad_serving/inline_content_ads/inline_content_ad_serving.cc

```diff
@@ -0,0 +1,129 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/ad_serving/inline_content_ads/inline_content_ad_serving.h"
+
+#include "base/rand_util.h"
+#include "bat/ads/ad_type.h"
+#include "bat/ads/inline_content_ad_info.h"
+#include "bat/ads/internal/ad_serving/ad_targeting/geographic/subdivision/subdivision_targeting.h"
+#include "bat/ads/internal/ad_targeting/ad_targeting.h"
+#include "bat/ads/internal/ad_targeting/ad_targeting_segment.h"
+#include "bat/ads/internal/ads/inline_content_ads/inline_content_ad_builder.h"
+#include "bat/ads/internal/bundle/creative_inline_content_ad_info.h"
+#include "bat/ads/internal/eligible_ads/inline_content_ads/eligible_inline_content_ads.h"
+#include "bat/ads/internal/features/inline_content_ads/inline_content_ads_features.h"
+#include "bat/ads/internal/logging.h"
+#include "bat/ads/internal/p2a/p2a_ad_opportunities/p2a_ad_opportunity.h"
+#include "bat/ads/internal/resources/frequency_capping/anti_targeting_resource.h"
+
+namespace ads {
+namespace inline_content_ads {
+
+AdServing::AdServing(
+    AdTargeting* ad_targeting,
+    ad_targeting::geographic::SubdivisionTargeting* subdivision_targeting,
+    resource::AntiTargeting* anti_targeting_resource)
+    : ad_targeting_(ad_targeting),
+      subdivision_targeting_(subdivision_targeting),
+      anti_targeting_resource_(anti_targeting_resource),
+      eligible_ads_(std::make_unique<EligibleAds>(subdivision_targeting,
+                                                  anti_targeting_resource)) {
+  DCHECK(ad_targeting_);
+  DCHECK(subdivision_targeting_);
+  DCHECK(anti_targeting_resource_);
+}
+
+AdServing::~AdServing() = default;
+
+void AdServing::AddObserver(InlineContentAdServingObserver* observer) {
+  DCHECK(observer);
+  observers_.AddObserver(observer);
+}
+
+void AdServing::RemoveObserver(InlineContentAdServingObserver* observer) {
+  DCHECK(observer);
+  observers_.RemoveObserver(observer);
+}
+
+void AdServing::MaybeServeAd(const std::string& size,
+                             GetInlineContentAdCallback callback) {
+  const SegmentList segments = ad_targeting_->GetSegments();
+
+  InlineContentAdInfo inline_content_ad;
+
+  if (!features::inline_content_ads::IsEnabled()) {
+    callback(/* success */ false, size, inline_content_ad);
+    return;
+  }
+
+  DCHECK(eligible_ads_);
+  eligible_ads_->GetForSegments(
+      segments, size,
+      [=](const bool was_allowed, const CreativeInlineContentAdList& ads) {
+        if (was_allowed) {
+          p2a::RecordAdOpportunityForSegments(AdType::kInlineContentAd,
```

Comment:
not sure we want to collate ad opportunity events across creative types. Maybe we keep it to ad notifications for now and re-think analytics in a separate ticket. cc @jsecretan

Reply:
Can we tell the types apart in P2A?  If not, might make sense like you said to just keep the opportunities focused on notifications and circle back

Reply:
Yeap, all types are separated

Reply:
Ok cool, then I'm good either way, happy to jettison anything analytics that might take longer, but sounds like it might not be necessary

Reply:
@jsecretan @tmancey we will not be able to tell type apart on the server side. So as not to conflate maybe we keep P2A only on ad notifications for now? I think we need to improve on P2A anyways so maybe we'll add support for types when we circle back?

Reply:
Sgtm @moritzhaller

Reply:
Discussed with @moritzhaller and all good